prov/efa/src/rxr/rxr_pkt_type_data.c

```diff
@@ -232,10 +232,9 @@ void rxr_pkt_handle_data_send_completion(struct rxr_ep *ep,
 	struct rxr_tx_entry *tx_entry;
 
 	tx_entry = (struct rxr_tx_entry *)pkt_entry->x_entry;
-	tx_entry->bytes_acked +=
-		rxr_get_data_pkt(pkt_entry->pkt)->hdr.seg_size;
-
-	if (tx_entry->total_len == tx_entry->bytes_acked)
+	tx_entry->bytes_acked += rxr_get_data_pkt(pkt_entry->pkt)->hdr.seg_size;
+	if (tx_entry->total_len == tx_entry->bytes_acked &&
+	    !(tx_entry->fi_flags & RXR_DELIVERY_COMPLETE_REQUESTED))
```

Comment:
I think we have a problem here, which is that the receipt might arrive before we got the completion of the last data packet. In which case, tx_entry might have been released, or have been allocated for other message. So if delivery complete is used, we cannot touch tx_entry. So we should check at the beginning of this function, if ep and peer both support DELIVERY_COMPLETE, we do not proceed (we do not even touch tx_entry).

Reply:
Current logic determines whether `rxr_cq_handle_tx_completion` is called in `rxr_pkt_handle_data_send_completion` or `rxr_pkt_handle_receipt_recv` determine which one is first called. If `rxr_pkt_handle_data_send_completion` is called first, then `rxr_cq_handle_tx_completion` will be called in `rxr_pkt_handle_receipt_recv` and vice versa.